util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+	if len(path.GetElem()) == 0 {
+		ResetIndent()
+		return []interface{}{root}, []*yang.Entry{schema}, nil
+	}
+	if schema == nil {
+		return nil, nil, fmt.Errorf("nil schema for data element type %T, remaining path %v", root, path)
+	}
+	// Strip off the absolute path prefix since the relative and absolute paths
+	// are assumed to be equal.
+	if path.GetElem()[0].GetName() == "" {
+		path.Elem = path.GetElem()[1:]
+	}
+
+	Indent()
+	DbgPrint("GetNode next path %v, value %v", path.GetElem()[0], ValueStr(root))
+
+	switch {
+	case schema.IsContainer() || (schema.IsList() && IsTypeStructPtr(reflect.TypeOf(root))):
+		// Either a container or list schema with struct data node (which could
+		// be an element of a list).
+		return getNodesContainer(schema, root, path)
+	case schema.IsList():
+		// A list schema with the list data node. Must find the element selected
+		// by the path.
+		return getNodesList(schema, root, path)
+	}
+
+	return nil, nil, fmt.Errorf("bad schema type for %s, struct type %T", schema.Name, root)
+}
+
+// getNodesContainer traverses the container root, which must be a struct ptr
+// type and matches each field against the first path element in path. If a
+// field matches, it recurses into that field with the remaining path.
+func getNodesContainer(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
+	DbgPrint("getNodesContainer: schema %s, next path %v, value %v", schema.Name, path.GetElem()[0], ValueStr(root))
+
+	rv := reflect.ValueOf(root)
+	if !IsValueStructPtr(rv) {
+		return nil, nil, fmt.Errorf("getNodesContainer: root has type %T, expect struct ptr", root)
+	}
+
+	v := rv.Elem()
+
+	for i := 0; i < v.NumField(); i++ {
+		f := v.Field(i)
+		ft := v.Type().Field(i)
+		cschema, err := FieldSchema(schema, ft)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error for schema for type %T, field name %s: %s", root, ft.Name, err)
+		}
+		if cschema == nil {
+			return nil, nil, fmt.Errorf("could not find schema for type %T, field name %s", root, ft.Name)
+		}
+		cschema, err = ResolveLeafRef(cschema)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error for schema for type %T, field name %s: %s", root, ft.Name, err)
+		}
+
+		DbgPrint("check field name %s", cschema.Name)
+		ps, err := SchemaPaths(ft)
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, p := range ps {
+			if pathMatchesPrefix(path, p) {
+				// don't trim whole prefix  for keyed list since name and key
+				// are a in the same element.
+				to := len(p)
+				if IsTypeMap(ft.Type) {
+					to--
+				}
+				return getNodesInternal(cschema, f.Interface(), trimGNMIPathPrefix(path, p[0:to]))
+			}
+		}
+	}
+
+	return nil, nil, DbgErr(fmt.Errorf("could not find path in tree beyond schema node %s, (type %T), remaining path %v", schema.Name, root, path))
+}
+
+// getNodesList traverses the list root, which must be a map of struct
+// type and matches each map key against the first path element in path. If the
```

Comment:
Against the keys specified in the first `PathElem` of the `Path`?

Reply:
sg, thanks.